app/src/main/java/io/github/trojan_gfw/igniter/settings/presenter/SettingsPresenter.java

```diff
@@ -0,0 +1,83 @@
+package io.github.trojan_gfw.igniter.settings.presenter;
+
+import androidx.annotation.NonNull;
+
+import java.util.List;
+import java.util.regex.Pattern;
+
+import io.github.trojan_gfw.igniter.common.os.Task;
+import io.github.trojan_gfw.igniter.common.os.Threads;
+import io.github.trojan_gfw.igniter.settings.contract.SettingsContract;
+import io.github.trojan_gfw.igniter.settings.data.ISettingsDataManager;
+
+public class SettingsPresenter implements SettingsContract.Presenter {
+    private final SettingsContract.View mView;
+    private final ISettingsDataManager mDataManager;
+    private boolean mHasModifiedDNS;
+
+    public SettingsPresenter(SettingsContract.View view, ISettingsDataManager dataManager) {
+        mView = view;
+        mDataManager = dataManager;
+        view.setPresenter(this);
+    }
+
+    @Override
+    public void saveDNSList(@NonNull List<String> dnsList) {
+        mView.showLoading();
+        Pattern pattern = Pattern.compile("^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$");
```

Comment:
It looks like this pattern supports IPv4 only.

Reply:
> It looks like this pattern supports IPv4 only.

Yes, is it necessary to support IPv6 now?